docs/source/pytest.rst

```diff
@@ -0,0 +1,435 @@
+
+    @pytest.mark.cocotb(timescale=("1ns", "1ps"))
+    def test_dut_using_different_timescale(sample_module: HDL) -> None:
+        """Test DUT using different timescale."""
+        sample_module.test()
+
+    @pytest.mark.cocotb(timeout=(200, "ns"))
+    async def test_dut_feature_with_timeout(dut) -> None:
+        """Test DUT feature. It must finish within 200 nanoseconds."""
+
+Additionally, positional arguments of :py:deco:`pytest.mark.cocotb` marker are equivalent to
+``test_module`` argument from :py:func:`cocotb.runner.Runner.test`.
+
+.. code:: python
+
+    @pytest.mark.cocotb("test_dut_tb_1", "test_dut_tb_2")
+    def test_dut_using_different_testbenches(sample_module: HDL) -> None:
+        """Use cocotb tests from ``test_dut_tb_1.py`` and ``test_dut_tb_2.py`` files to test DUT."""
+        sample_module.test()
+
+If no positional arguments were provided to :py:deco:`pytest.mark.cocotb`,
+plugin will load current Python module where :py:deco:`pytest.mark.cocotb` was used as cocotb testbench (Python file with
+cocotb tests).
+
+.. code:: python
+
+    @pytest.mark.cocotb
+    def test_dut_using_default_testbench(sample_module: HDL) -> None:
+        """Test DUT with cocotb tests defined in the same Python file as this test function."""
+        sample_module.test()
+
+    async def test_dut_feature_1(dut) -> None:
+        """Test DUT feature 1."""
+
+If ``toplevel`` argument is empty/non-set, plugin will use name of first test module but without
+``test_*`` prefix or ``*_test`` suffix. For example, if test module was ``test_dut`` then
+name of HDL top level design will be ``dut``.
+
+.. code:: python
+
+    @pytest.mark.cocotb
+    def test_dut_using_default_toplevel(sample_module: HDL) -> None:
+        """Test DUT with default top level associated with name of test file as this test function."""
+        sample_module.test()
+
+    @pytest.mark.cocotb(toplevel="sample_submodule")
+    def test_dut_using_different_toplevel(sample_module: HDL) -> None:
+        """Test DUT with different top level that was set at fixture level."""
+        sample_module.test()
+
+Using the :py:deco:`pytest.mark.cocotb` marker is optional for test functions if they meet the following criteria:
+
+* start with ``test_``
+* is a coroutine function (``async def``)
+* have the first positional argument name match the ``toplevel`` (such as ``dut``)
+
+Non-``async`` functions marked with :py:deco:`pytest.mark.cocotb` are control functions run by pytest.
+They can run simulations by invoking :py:func:`cocotb_tools.pytest.hdl.HDL.test`
+or :py:func:`cocotb_tools.runner.Runner.test`.
+
+.. code:: python
+
+    import pytest
+    from cocotb_tools.pytest.hdl import HDL
+
+
+    # First, define new HDL design, add HDL source files to it and build it
+    @pytest.fixture(name="sample_module")
+    def sample_module_fixture(hdl: HDL) -> HDL:
+        """Define new HDL design by adding HDL source files to it."""
+        hdl.toplevel = "sample_module"
```

Comment:
So because `toplevel` is set, don't the async functions below need to use `sample_module` for their first position argument and not `dut`?

Also, does it work if there are 2+ HDL fixtures in the same test module with different tests assigned to each based on the bulleted criteria list above?

Reply:
@marlonjames I think it is worthy to put these above examples in `pytest.rst` documentation to make it clear :thinking:

Reply:
@marlonjames Thanks for pointing this out! You are correct. I have applied this change in this commit 7d5e640fd9ab20bdd993ebda06fe3a4288da6e28